refactor: complex voucher tools

jniles opened the pull request:
This PR tracks work towards a large-scale refactor of Voucher Tools.  Each Voucher Tool is being revisited and rewritten to increase performance and address any reported bugs.  Duplicated code is being consolidated into shared services.

~~**Note** - This work is still ongoing and should not be merged yet.~~  This can now be merged anytime.

**Major Changes**
 1. Replace bhBreadcrumb with a simple headercrumb.  This change reduces the JS code by about 25 lines.  It also makes the voucher tools much easier to update, re-order, and change.  See #1923.
 2. Move shared code to the `VoucherToolKitService`.  This is particular evident with the `generateRow()` function that was used identically in ever controller.
 3. All forms that require accounts now use the `bhAccountSelect` component instead of manually downloading accounts themselves.
 4. Update Prise En Charge text.  Fixes #1956.
 5. Ensure modal ui-grids function correctly and more efficiently.
   a. Cost/balance sorting is now correct.
   b. Remove unnecessary cell templates (e.g. `balance.cell.html`)
   c. Consolidate `onSelectionChanged` listeners - reduces number of listeners by 50%.
   d. Consolidate aggregation functions into `aggregate()`.  No longer defines a new function on every selection change event.


-----
Thank you for contributing!

Before submitting this pull request, please verify that you have:
 - [x] Run your code through eslint.
 - [x] Run integration tests.
 - [x] Run end-to-end tests.
 - [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the online review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

jniles committed:
refactor(vouchers): remove bhBreadcrumb bar

This commit removes the bhBreadcrumb header bar from the complex voucher
page, which should make it easier to modify complex voucher tools in the
future.

References #1923.

jniles committed:
fix(vouchers): rewrite convention payments

This commit fixes the voucher tool for convention payments.  The
previous version downloaded the entirety of the Debtors table and
Invoices table.  The new version only downloads data that it needs.  It
also fixes a few usability issues such as sorting by cost.

jniles committed:
refactor(vouchers): rewrite cash transfer tool

This commit guts the cash transfer tool, rewriting it to use the
bhAccountSelection component and bhCurrencyInput component.  This
drastically reduces and compartmentalizes the code.  It also removes
the requirement on downloading lots of extraneous data to serve the
page.

jniles committed:
refactor(vouchers): mv row generation into service

This commit consolidates all the `generateRow()` functions into the
Voucher ToolKit Service.  These were essentially all duplicated code -
by consolidating them into the service cleans up each controller by
10 lines of code or so at the cost of a single import.

jniles committed:
refactor(vouchers): generic income/expense tools

This commit refactors the code in the income and expense tools to be
more compact.  It also leverages existing components to reduce the
amount of code in the parent controller.

jniles committed:
fix(vouchers): rework prise en charge

This commit cleans up the prise en charge voucher tool with refactors
from other modules.  It additionally adds in human readable description
text automatically on tool submission.

Closes #1956.

jniles committed:
test(voucher): update end to end tests

This commit updates the validation and end to end tests across the
voucher tool forms.

mbayopanda commented on client/src/modules/vouchers/complex-voucher.ctrl.js:
```diff
@@ -76,32 +61,31 @@ function ComplexJournalVoucherController(Vouchers, $translate, Currencies, Sessi
+  vm.openConventionPaymentModal = function openConventionPaymentModal() {
+    Toolkit.openConventionPaymentModal()
+      .then(processVoucherToolRows);
+  };
```
👍

mbayopanda commented on client/src/modules/vouchers/complex-voucher.ctrl.js:
```diff
@@ -45,22 +46,6 @@ function ComplexJournalVoucherController(Vouchers, $translate, Currencies, Sessi
-    icon   : 'fa-cogs',
-    option : Toolkit.options,
-  }];
-
```
💃 the mystery of the black box of `breadcrumb component` dynamic tools is now exposed in a clear way.

mbayopanda commented on client/src/modules/vouchers/toolkit/cash_transfer/cash_transfer.html:
```diff
@@ -37,41 +37,19 @@
+      on-select-callback="ToolCtrl.onSelectAccountCallback(account)"
+      exclude-title-accounts="true">
+      (<label class="control-label" translate>FORM.LABELS.CREDIT</label>)
+    </bh-account-select>
```
👍 Ooura... Component

jniles replied:
[Ooura](http://www.dictionary.com/misspelling?term=ooura&s=t) :laughing:

mbayopanda reviewed (commented):
The code LGTM

jniles committed:
deps: roll back ui-select updates

This commit fixes issues with the ui-select dropdown on certain fields.
It turns out the newest version does not always drop down, severely
limiting a user's ability to work.

jniles committed:
fix(bhTransactionTypes): default to empty array

The bhTransactionTypes will now default to an empty array if the
transactions types are not provided to it.  This makes the transaction
types binding truly optional.  A number of small issues have also been
cleaned up: addition of `ngForm` to make sure the `ngModelOptions` is
respected, code updated to standards.
